# Pre-Work

In [1]:
import warnings
warnings.filterwarnings("ignore", category=FutureWarning)
import tensorflow as tf
import tensorflow.keras as keras
from tensorflow.keras import Sequential
from tensorflow.keras.layers import Conv2D, Dense, Flatten, MaxPool2D, Dropout, BatchNormalization
from tensorflow.keras.callbacks import ModelCheckpoint

In [2]:
gpus = tf.config.experimental.list_physical_devices('GPU')
if gpus:
    # Restrict TensorFlow to only allocate 1GB * 2 of memory on the first GPU
    try:
        tf.config.experimental.set_virtual_device_configuration(
            gpus[0],
            [tf.config.experimental.VirtualDeviceConfiguration(memory_limit=1024 * 2)])
        logical_gpus = tf.config.experimental.list_logical_devices('GPU')
        print(len(gpus), "Physical GPUs,", len(logical_gpus), "Logical GPUs")
    except RuntimeError as e:
        # Virtual devices must be set before GPUs have been initialized
        print(e)

1 Physical GPUs, 1 Logical GPUs


# Experiment 1 - Initial Try

### Create Weight Checkpoint

In [47]:
checkpoint = ModelCheckpoint('weights.hdf5', \
                              mode='max', \
                              monitor='val_accuracy', \
                              save_best_only=True)
callbacks_list=[checkpoint]

### Datagen

In [48]:
trainpath = '../data/raw/imgs/train'
testpath = '../data/raw/imgs/testlabeled'

In [49]:
datagen = keras.preprocessing.image.ImageDataGenerator()
testdatagen = keras.preprocessing.image.ImageDataGenerator()

In [50]:
train = datagen.flow_from_directory(trainpath)

Found 22424 images belonging to 10 classes.


In [51]:
test = datagen.flow_from_directory(testpath)

Found 200 images belonging to 10 classes.


### Build Model

In [52]:
model = Sequential()

In [53]:
model.add(Conv2D(10, kernel_size=3, activation='relu', input_shape=(256, 256, 3)))

In [54]:
model.add(MaxPool2D(10))

In [55]:
model.add(Flatten())

In [56]:
model.add(Dense(10, activation='softmax'))

In [57]:
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

In [58]:
model.summary()

Model: "sequential_3"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_3 (Conv2D)            (None, 254, 254, 10)      280       
_________________________________________________________________
max_pooling2d_3 (MaxPooling2 (None, 25, 25, 10)        0         
_________________________________________________________________
flatten_3 (Flatten)          (None, 6250)              0         
_________________________________________________________________
dense_3 (Dense)              (None, 10)                62510     
Total params: 62,790
Trainable params: 62,790
Non-trainable params: 0
_________________________________________________________________


### Fit Model

In [59]:
model.fit(train, \
          epochs=20, \
          steps_per_epoch=40, \
          validation_data=test, \
          callbacks=callbacks_list)

  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train for 40 steps, validate for 7 steps
Epoch 1/20
40/40 [==============================] - 14s 343ms/step - loss: 206.4199 - accuracy: 0.1375 - val_loss: 61.2814 - val_accuracy: 0.1950
Epoch 2/20
40/40 [==============================] - 12s 295ms/step - loss: 25.8079 - accuracy: 0.3070 - val_loss: 25.6535 - val_accuracy: 0.2300
Epoch 3/20
40/40 [==============================] - 13s 331ms/step - loss: 9.2154 - accuracy: 0.4875 - val_loss: 24.1452 - val_accuracy: 0.2300
Epoch 4/20
40/40 [==============================] - 13s 328ms/step - loss: 5.7722 - accuracy: 0.6305 - val_loss: 24.4043 - val_accuracy: 0.2550
Epoch 5/20
40/40 [==============================] - 12s 293ms/step - loss: 3.1833 - accuracy: 0.7484 - val_loss: 23.5417 - val_accuracy: 0.2450
Epoch 6/20
40/40 [==============================] - 12s 298ms/step - loss: 2.6024 - accuracy: 0.8050 - val_loss: 22.6050 - val_accuracy: 0.2700
Epoch 7/20
40/40 [========================

Overfit on the training data, did not generalize to the test sample

# Experiment 2 - Increase Kernel Size

### Create Weight Checkpoint

In [60]:
checkpoint = ModelCheckpoint('weights.hdf5', \
                              mode='max', \
                              monitor='val_accuracy', \
                              save_best_only=True)
callbacks_list=[checkpoint]

### Datagen

In [61]:
trainpath = '../data/raw/imgs/train'
testpath = '../data/raw/imgs/testlabeled'

In [62]:
datagen = keras.preprocessing.image.ImageDataGenerator()
testdatagen = keras.preprocessing.image.ImageDataGenerator()

In [63]:
train = datagen.flow_from_directory(trainpath)

Found 22424 images belonging to 10 classes.


In [64]:
test = datagen.flow_from_directory(testpath)

Found 200 images belonging to 10 classes.


### Build Model

In [65]:
model = Sequential()

In [66]:
model.add(Conv2D(10, kernel_size=5, activation='relu', input_shape=(256, 256, 3)))

In [67]:
model.add(MaxPool2D(10))

In [68]:
model.add(Flatten())

In [69]:
model.add(Dense(10, activation='softmax'))

In [70]:
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

In [71]:
model.summary()

Model: "sequential_4"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_4 (Conv2D)            (None, 252, 252, 10)      760       
_________________________________________________________________
max_pooling2d_4 (MaxPooling2 (None, 25, 25, 10)        0         
_________________________________________________________________
flatten_4 (Flatten)          (None, 6250)              0         
_________________________________________________________________
dense_4 (Dense)              (None, 10)                62510     
Total params: 63,270
Trainable params: 63,270
Non-trainable params: 0
_________________________________________________________________


### Fit Model

In [72]:
model.fit(train, \
          epochs=20, \
          steps_per_epoch=40, \
          validation_data=test, \
          callbacks=callbacks_list)

  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train for 40 steps, validate for 7 steps
Epoch 1/20
40/40 [==============================] - 14s 358ms/step - loss: 91.8692 - accuracy: 0.1266 - val_loss: 10.7892 - val_accuracy: 0.1200
Epoch 2/20
40/40 [==============================] - 13s 329ms/step - loss: 5.1260 - accuracy: 0.1937 - val_loss: 6.1438 - val_accuracy: 0.1200
Epoch 3/20
40/40 [==============================] - 13s 323ms/step - loss: 3.0122 - accuracy: 0.2820 - val_loss: 5.3370 - val_accuracy: 0.1150
Epoch 4/20
40/40 [==============================] - 12s 306ms/step - loss: 2.2932 - accuracy: 0.3602 - val_loss: 5.0220 - val_accuracy: 0.0950
Epoch 5/20
40/40 [==============================] - 12s 306ms/step - loss: 1.8092 - accuracy: 0.4742 - val_loss: 4.7914 - val_accuracy: 0.1100
Epoch 6/20
40/40 [==============================] - 11s 285ms/step - loss: 1.3916 - accuracy: 0.5734 - val_loss: 4.9906 - val_accuracy: 0.1000
Epoch 7/20
40/40 [==============================]

Larger kernel performed worse. Let's try a smaller kernel.

# Experiment 3 - Decrease Kernel Size

### Create Weight Checkpoint

In [73]:
checkpoint = ModelCheckpoint('weights.hdf5', \
                              mode='max', \
                              monitor='val_accuracy', \
                              save_best_only=True)
callbacks_list=[checkpoint]

### Datagen

In [74]:
trainpath = '../data/raw/imgs/train'
testpath = '../data/raw/imgs/testlabeled'

In [75]:
datagen = keras.preprocessing.image.ImageDataGenerator()
testdatagen = keras.preprocessing.image.ImageDataGenerator()

In [76]:
train = datagen.flow_from_directory(trainpath)

Found 22424 images belonging to 10 classes.


In [77]:
test = datagen.flow_from_directory(testpath)

Found 200 images belonging to 10 classes.


### Build Model

In [78]:
model = Sequential()

In [79]:
model.add(Conv2D(10, kernel_size=2, activation='relu', input_shape=(256, 256, 3)))

In [80]:
model.add(MaxPool2D(10))

In [81]:
model.add(Flatten())

In [82]:
model.add(Dense(10, activation='softmax'))

In [83]:
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

In [84]:
model.summary()

Model: "sequential_5"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_5 (Conv2D)            (None, 255, 255, 10)      130       
_________________________________________________________________
max_pooling2d_5 (MaxPooling2 (None, 25, 25, 10)        0         
_________________________________________________________________
flatten_5 (Flatten)          (None, 6250)              0         
_________________________________________________________________
dense_5 (Dense)              (None, 10)                62510     
Total params: 62,640
Trainable params: 62,640
Non-trainable params: 0
_________________________________________________________________


### Fit Model

In [85]:
model.fit(train, \
          epochs=20, \
          steps_per_epoch=40, \
          validation_data=test, \
          callbacks=callbacks_list)

  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train for 40 steps, validate for 7 steps
Epoch 1/20
40/40 [==============================] - 11s 269ms/step - loss: 189.8808 - accuracy: 0.1531 - val_loss: 73.4405 - val_accuracy: 0.1400
Epoch 2/20
40/40 [==============================] - 10s 248ms/step - loss: 25.9073 - accuracy: 0.3313 - val_loss: 26.9899 - val_accuracy: 0.1800
Epoch 3/20
40/40 [==============================] - 10s 259ms/step - loss: 10.4909 - accuracy: 0.4836 - val_loss: 27.0480 - val_accuracy: 0.2450
Epoch 4/20
40/40 [==============================] - 10s 259ms/step - loss: 8.1029 - accuracy: 0.6031 - val_loss: 26.4237 - val_accuracy: 0.2300
Epoch 5/20
40/40 [==============================] - 11s 265ms/step - loss: 4.3869 - accuracy: 0.7258 - val_loss: 26.1492 - val_accuracy: 0.2600
Epoch 6/20
40/40 [==============================] - 10s 248ms/step - loss: 2.2677 - accuracy: 0.8141 - val_loss: 25.8212 - val_accuracy: 0.2650
Epoch 7/20
40/40 [=======================

Kernel_size=3 performed best so far

# Experiment 4 - Add Dropout Layer

### Create Weight Checkpoint

In [94]:
checkpoint = ModelCheckpoint('weights.hdf5', \
                              mode='max', \
                              monitor='val_accuracy', \
                              save_best_only=True)
callbacks_list=[checkpoint]

### Datagen

In [95]:
trainpath = '../data/raw/imgs/train'
testpath = '../data/raw/imgs/testlabeled'

In [96]:
datagen = keras.preprocessing.image.ImageDataGenerator()
testdatagen = keras.preprocessing.image.ImageDataGenerator()

In [97]:
train = datagen.flow_from_directory(trainpath)

Found 22424 images belonging to 10 classes.


In [98]:
test = datagen.flow_from_directory(testpath)

Found 200 images belonging to 10 classes.


### Build Model

In [99]:
model = Sequential()

In [100]:
model.add(Conv2D(10, kernel_size=3, activation='relu', input_shape=(256, 256, 3)))

In [101]:
model.add(Dropout(0.25))

In [102]:
model.add(MaxPool2D(10))

In [103]:
model.add(Flatten())

In [104]:
model.add(Dense(10, activation='softmax'))

In [105]:
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

In [106]:
model.summary()

Model: "sequential_7"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_6 (Conv2D)            (None, 254, 254, 10)      280       
_________________________________________________________________
dropout_1 (Dropout)          (None, 254, 254, 10)      0         
_________________________________________________________________
max_pooling2d_7 (MaxPooling2 (None, 25, 25, 10)        0         
_________________________________________________________________
flatten_6 (Flatten)          (None, 6250)              0         
_________________________________________________________________
dense_6 (Dense)              (None, 10)                62510     
Total params: 62,790
Trainable params: 62,790
Non-trainable params: 0
_________________________________________________________________


### Fit Model

In [107]:
model.fit(train, \
          epochs=20, \
          steps_per_epoch=40, \
          validation_data=test, \
          callbacks=callbacks_list)

  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train for 40 steps, validate for 7 steps
Epoch 1/20
40/40 [==============================] - 10s 247ms/step - loss: 49.5038 - accuracy: 0.1656 - val_loss: 17.6243 - val_accuracy: 0.1500
Epoch 2/20
40/40 [==============================] - 9s 232ms/step - loss: 10.2765 - accuracy: 0.3523 - val_loss: 12.6249 - val_accuracy: 0.2800
Epoch 3/20
40/40 [==============================] - 9s 231ms/step - loss: 4.7453 - accuracy: 0.5672 - val_loss: 12.1243 - val_accuracy: 0.2750
Epoch 4/20
40/40 [==============================] - 10s 238ms/step - loss: 2.6323 - accuracy: 0.7195 - val_loss: 12.2602 - val_accuracy: 0.3050
Epoch 5/20
40/40 [==============================] - 9s 232ms/step - loss: 1.7278 - accuracy: 0.7859 - val_loss: 12.7241 - val_accuracy: 0.3000
Epoch 6/20
40/40 [==============================] - 9s 233ms/step - loss: 1.1139 - accuracy: 0.8492 - val_loss: 12.9153 - val_accuracy: 0.3100
Epoch 7/20
40/40 [=============================

Did a little better

# Experiment 5 - Remove MaxPooling Layer

### Create Weight Checkpoint

In [108]:
checkpoint = ModelCheckpoint('weights.hdf5', \
                              mode='max', \
                              monitor='val_accuracy', \
                              save_best_only=True)
callbacks_list=[checkpoint]

### Datagen

In [109]:
trainpath = '../data/raw/imgs/train'
testpath = '../data/raw/imgs/testlabeled'

In [110]:
datagen = keras.preprocessing.image.ImageDataGenerator()
testdatagen = keras.preprocessing.image.ImageDataGenerator()

In [111]:
train = datagen.flow_from_directory(trainpath)

Found 22424 images belonging to 10 classes.


In [112]:
test = datagen.flow_from_directory(testpath)

Found 200 images belonging to 10 classes.


### Build Model

In [113]:
model = Sequential()

In [114]:
model.add(Conv2D(10, kernel_size=3, activation='relu', input_shape=(256, 256, 3)))

In [115]:
model.add(Dropout(0.25))

In [116]:
model.add(Flatten())

In [117]:
model.add(Dense(10, activation='softmax'))

In [118]:
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

In [119]:
model.summary()

Model: "sequential_8"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_7 (Conv2D)            (None, 254, 254, 10)      280       
_________________________________________________________________
dropout_2 (Dropout)          (None, 254, 254, 10)      0         
_________________________________________________________________
flatten_7 (Flatten)          (None, 645160)            0         
_________________________________________________________________
dense_7 (Dense)              (None, 10)                6451610   
Total params: 6,451,890
Trainable params: 6,451,890
Non-trainable params: 0
_________________________________________________________________


### Fit Model

In [120]:
model.fit(train, \
          epochs=20, \
          steps_per_epoch=40, \
          validation_data=test, \
          callbacks=callbacks_list)

  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train for 40 steps, validate for 7 steps
Epoch 1/20
40/40 [==============================] - 10s 262ms/step - loss: 11464.7284 - accuracy: 0.1328 - val_loss: 1047.1991 - val_accuracy: 0.1400
Epoch 2/20
40/40 [==============================] - 10s 258ms/step - loss: 326.8999 - accuracy: 0.4602 - val_loss: 142.2167 - val_accuracy: 0.3550
Epoch 3/20
40/40 [==============================] - 9s 226ms/step - loss: 18.1810 - accuracy: 0.8375 - val_loss: 116.1318 - val_accuracy: 0.3450
Epoch 4/20
40/40 [==============================] - 9s 232ms/step - loss: 6.6618 - accuracy: 0.9039 - val_loss: 124.9605 - val_accuracy: 0.3200
Epoch 5/20
40/40 [==============================] - 9s 226ms/step - loss: 8.2324 - accuracy: 0.9000 - val_loss: 137.4872 - val_accuracy: 0.3300
Epoch 6/20
40/40 [==============================] - 9s 231ms/step - loss: 5.2438 - accuracy: 0.9266 - val_loss: 119.1479 - val_accuracy: 0.3150
Epoch 7/20
40/40 [=================

About the same with way more parameters. Still overfitting like crazy.

# Experiment 6 - Replace Dropout with BatchNormalization

### Create Weight Checkpoint

In [134]:
checkpoint = ModelCheckpoint('weights.hdf5', \
                              mode='max', \
                              monitor='val_accuracy', \
                              save_best_only=True)
callbacks_list=[checkpoint]

### Datagen

In [135]:
trainpath = '../data/raw/imgs/train'
testpath = '../data/raw/imgs/testlabeled'

In [136]:
datagen = keras.preprocessing.image.ImageDataGenerator()
testdatagen = keras.preprocessing.image.ImageDataGenerator()

In [137]:
train = datagen.flow_from_directory(trainpath)

Found 22424 images belonging to 10 classes.


In [138]:
test = datagen.flow_from_directory(testpath)

Found 200 images belonging to 10 classes.


### Build Model

In [139]:
model = Sequential()

In [140]:
model.add(Conv2D(10, kernel_size=3, activation='relu', input_shape=(256, 256, 3)))

In [141]:
model.add(BatchNormalization())

In [142]:
model.add(Flatten())

In [143]:
model.add(Dense(10, activation='softmax'))

In [144]:
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

In [145]:
model.summary()

Model: "sequential_10"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_9 (Conv2D)            (None, 254, 254, 10)      280       
_________________________________________________________________
batch_normalization (BatchNo (None, 254, 254, 10)      40        
_________________________________________________________________
flatten_9 (Flatten)          (None, 645160)            0         
_________________________________________________________________
dense_9 (Dense)              (None, 10)                6451610   
Total params: 6,451,930
Trainable params: 6,451,910
Non-trainable params: 20
_________________________________________________________________


### Fit Model

In [146]:
model.fit(train, \
          epochs=20, \
          steps_per_epoch=40, \
          validation_data=test, \
          callbacks=callbacks_list)

  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train for 40 steps, validate for 7 steps
Epoch 1/20
40/40 [==============================] - 10s 257ms/step - loss: 42.8838 - accuracy: 0.3375 - val_loss: 26.2029 - val_accuracy: 0.2200
Epoch 2/20
40/40 [==============================] - 9s 232ms/step - loss: 7.1125 - accuracy: 0.6984 - val_loss: 31.2764 - val_accuracy: 0.2950
Epoch 3/20
40/40 [==============================] - 9s 237ms/step - loss: 3.9108 - accuracy: 0.8336 - val_loss: 26.1086 - val_accuracy: 0.3000
Epoch 4/20
40/40 [==============================] - 9s 233ms/step - loss: 2.4493 - accuracy: 0.9000 - val_loss: 33.3006 - val_accuracy: 0.3200
Epoch 5/20
40/40 [==============================] - 9s 225ms/step - loss: 2.6187 - accuracy: 0.8969 - val_loss: 33.1035 - val_accuracy: 0.2350
Epoch 6/20
40/40 [==============================] - 9s 236ms/step - loss: 1.9513 - accuracy: 0.9211 - val_loss: 38.5788 - val_accuracy: 0.3400
Epoch 7/20
40/40 [==============================]

No better

# Experiment 7 - Bring back dropout, add a second convolutional layer

### Create Weight Checkpoint

In [162]:
checkpoint = ModelCheckpoint('weights.hdf5', \
                              mode='max', \
                              monitor='val_accuracy', \
                              save_best_only=True)
callbacks_list=[checkpoint]

### Datagen

In [163]:
trainpath = '../data/raw/imgs/train'
testpath = '../data/raw/imgs/testlabeled'

In [164]:
datagen = keras.preprocessing.image.ImageDataGenerator()
testdatagen = keras.preprocessing.image.ImageDataGenerator()

In [165]:
train = datagen.flow_from_directory(trainpath)

Found 22424 images belonging to 10 classes.


In [166]:
test = datagen.flow_from_directory(testpath)

Found 200 images belonging to 10 classes.


### Build Model

In [167]:
model = Sequential()

In [168]:
model.add(Conv2D(10, kernel_size=3, activation='relu', input_shape=(256, 256, 3)))

In [169]:
model.add(Dropout(0.25))

In [170]:
model.add(Conv2D(10, kernel_size=3, activation='relu'))

In [171]:
model.add(Dropout(0.25))

In [172]:
model.add(Flatten())

In [173]:
model.add(Dense(10, activation='softmax'))

In [174]:
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

In [175]:
model.summary()

Model: "sequential_12"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_12 (Conv2D)           (None, 254, 254, 10)      280       
_________________________________________________________________
dropout_5 (Dropout)          (None, 254, 254, 10)      0         
_________________________________________________________________
conv2d_13 (Conv2D)           (None, 252, 252, 10)      910       
_________________________________________________________________
dropout_6 (Dropout)          (None, 252, 252, 10)      0         
_________________________________________________________________
flatten_11 (Flatten)         (None, 635040)            0         
_________________________________________________________________
dense_11 (Dense)             (None, 10)                6350410   
Total params: 6,351,600
Trainable params: 6,351,600
Non-trainable params: 0
___________________________________________

### Fit Model

In [176]:
model.fit(train, \
          epochs=20, \
          steps_per_epoch=40, \
          validation_data=test, \
          callbacks=callbacks_list)

  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train for 40 steps, validate for 7 steps
Epoch 1/20
40/40 [==============================] - 11s 263ms/step - loss: 3991.8533 - accuracy: 0.1289 - val_loss: 51.7726 - val_accuracy: 0.1400
Epoch 2/20
40/40 [==============================] - 9s 236ms/step - loss: 27.9060 - accuracy: 0.3680 - val_loss: 6.1529 - val_accuracy: 0.1800
Epoch 3/20
40/40 [==============================] - 9s 234ms/step - loss: 2.3160 - accuracy: 0.6305 - val_loss: 4.7680 - val_accuracy: 0.1900
Epoch 4/20
40/40 [==============================] - 9s 236ms/step - loss: 1.2600 - accuracy: 0.7414 - val_loss: 4.1564 - val_accuracy: 0.2100
Epoch 5/20
40/40 [==============================] - 9s 227ms/step - loss: 0.9547 - accuracy: 0.7719 - val_loss: 4.1388 - val_accuracy: 0.1850
Epoch 6/20
40/40 [==============================] - 9s 227ms/step - loss: 0.7604 - accuracy: 0.8219 - val_loss: 4.0299 - val_accuracy: 0.1800
Epoch 7/20
40/40 [==============================] -

Performed worse

# Experiment 8 - Increase number of nodes in first convolutional layer

### Create Weight Checkpoint

In [177]:
checkpoint = ModelCheckpoint('weights.hdf5', \
                              mode='max', \
                              monitor='val_accuracy', \
                              save_best_only=True)
callbacks_list=[checkpoint]

### Datagen

In [178]:
trainpath = '../data/raw/imgs/train'
testpath = '../data/raw/imgs/testlabeled'

In [179]:
datagen = keras.preprocessing.image.ImageDataGenerator()
testdatagen = keras.preprocessing.image.ImageDataGenerator()

In [180]:
train = datagen.flow_from_directory(trainpath)

Found 22424 images belonging to 10 classes.


In [181]:
test = datagen.flow_from_directory(testpath)

Found 200 images belonging to 10 classes.


### Build Model

In [182]:
model = Sequential()

In [183]:
model.add(Conv2D(25, kernel_size=3, activation='relu', input_shape=(256, 256, 3)))

In [184]:
model.add(Dropout(0.25))

In [185]:
model.add(Conv2D(10, kernel_size=3, activation='relu'))

In [186]:
model.add(Dropout(0.25))

In [187]:
model.add(Flatten())

In [188]:
model.add(Dense(10, activation='softmax'))

In [189]:
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

In [190]:
model.summary()

Model: "sequential_13"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_14 (Conv2D)           (None, 254, 254, 25)      700       
_________________________________________________________________
dropout_7 (Dropout)          (None, 254, 254, 25)      0         
_________________________________________________________________
conv2d_15 (Conv2D)           (None, 252, 252, 10)      2260      
_________________________________________________________________
dropout_8 (Dropout)          (None, 252, 252, 10)      0         
_________________________________________________________________
flatten_12 (Flatten)         (None, 635040)            0         
_________________________________________________________________
dense_12 (Dense)             (None, 10)                6350410   
Total params: 6,353,370
Trainable params: 6,353,370
Non-trainable params: 0
___________________________________________

### Fit Model

In [191]:
model.fit(train, \
          epochs=20, \
          steps_per_epoch=40, \
          validation_data=test, \
          callbacks=callbacks_list)

  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train for 40 steps, validate for 7 steps
Epoch 1/20
40/40 [==============================] - 10s 257ms/step - loss: 585.7035 - accuracy: 0.1375 - val_loss: 2.3104 - val_accuracy: 0.1250
Epoch 2/20
40/40 [==============================] - 9s 226ms/step - loss: 2.3322 - accuracy: 0.1703 - val_loss: 2.3022 - val_accuracy: 0.1000
Epoch 3/20
40/40 [==============================] - 9s 225ms/step - loss: 2.2763 - accuracy: 0.1680 - val_loss: 2.3076 - val_accuracy: 0.1100
Epoch 4/20
40/40 [==============================] - 9s 225ms/step - loss: 2.1862 - accuracy: 0.2258 - val_loss: 2.3655 - val_accuracy: 0.0900
Epoch 5/20
40/40 [==============================] - 9s 225ms/step - loss: 1.9212 - accuracy: 0.3539 - val_loss: 2.6162 - val_accuracy: 0.0800
Epoch 6/20
40/40 [==============================] - 9s 225ms/step - loss: 1.6729 - accuracy: 0.4672 - val_loss: 2.7381 - val_accuracy: 0.1000
Epoch 7/20
40/40 [==============================] - 9s

Performed worse

# Experiment 9 - Add more Conv, Dense, Pooling, and Dropout Layers

### Create Weight Checkpoint

In [197]:
checkpoint = ModelCheckpoint('weights.hdf5', \
                              mode='max', \
                              monitor='val_accuracy', \
                              save_best_only=True)
callbacks_list=[checkpoint]

### Datagen

In [198]:
trainpath = '../data/raw/imgs/train'
testpath = '../data/raw/imgs/testlabeled'

In [199]:
datagen = keras.preprocessing.image.ImageDataGenerator()
testdatagen = keras.preprocessing.image.ImageDataGenerator()

In [200]:
train = datagen.flow_from_directory(trainpath)

Found 22424 images belonging to 10 classes.


In [201]:
test = datagen.flow_from_directory(testpath)

Found 200 images belonging to 10 classes.


### Build Model

In [202]:
model = Sequential()

In [203]:
model.add(Conv2D(10, kernel_size=3, activation='relu', input_shape=(256, 256, 3)))

In [204]:
model.add(Conv2D(10, kernel_size=3, activation='relu'))

In [205]:
model.add(MaxPool2D(10))

In [206]:
model.add(Conv2D(10, kernel_size=3, activation='relu'))

In [207]:
model.add(Conv2D(10, kernel_size=3, activation='relu'))

In [208]:
model.add(MaxPool2D(10))

In [209]:
model.add(Flatten())

In [211]:
model.add(Dense(25, activation='relu'))

In [212]:
model.add(Dropout(0.25))

In [213]:
model.add(Dense(10, activation='softmax'))

In [214]:
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

In [215]:
model.summary()

Model: "sequential_14"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_16 (Conv2D)           (None, 254, 254, 10)      280       
_________________________________________________________________
conv2d_17 (Conv2D)           (None, 252, 252, 10)      910       
_________________________________________________________________
max_pooling2d_8 (MaxPooling2 (None, 25, 25, 10)        0         
_________________________________________________________________
conv2d_18 (Conv2D)           (None, 23, 23, 10)        910       
_________________________________________________________________
conv2d_19 (Conv2D)           (None, 21, 21, 10)        910       
_________________________________________________________________
max_pooling2d_9 (MaxPooling2 (None, 2, 2, 10)          0         
_________________________________________________________________
flatten_13 (Flatten)         (None, 40)              

### Fit Model

In [216]:
model.fit(train, \
          epochs=20, \
          steps_per_epoch=40, \
          validation_data=test, \
          callbacks=callbacks_list)

  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train for 40 steps, validate for 7 steps
Epoch 1/20
40/40 [==============================] - 10s 253ms/step - loss: 10.0287 - accuracy: 0.1063 - val_loss: 2.4930 - val_accuracy: 0.1100
Epoch 2/20
40/40 [==============================] - 9s 226ms/step - loss: 2.3948 - accuracy: 0.1156 - val_loss: 2.3390 - val_accuracy: 0.0950
Epoch 3/20
40/40 [==============================] - 9s 224ms/step - loss: 2.3280 - accuracy: 0.1086 - val_loss: 2.3230 - val_accuracy: 0.1050
Epoch 4/20
40/40 [==============================] - 9s 230ms/step - loss: 2.3073 - accuracy: 0.1234 - val_loss: 2.3341 - val_accuracy: 0.1350
Epoch 5/20
40/40 [==============================] - 9s 225ms/step - loss: 2.3051 - accuracy: 0.1250 - val_loss: 2.3159 - val_accuracy: 0.1150
Epoch 6/20
40/40 [==============================] - 9s 231ms/step - loss: 2.3021 - accuracy: 0.1055 - val_loss: 2.3016 - val_accuracy: 0.1400
Epoch 7/20
40/40 [==============================] - 9s 

Performed worse

# Experiment 10 - AlexNet Architecture

### Create Weight Checkpoint

In [3]:
checkpoint = ModelCheckpoint('weights.hdf5', \
                              mode='max', \
                              monitor='val_accuracy', \
                              save_best_only=True)
callbacks_list=[checkpoint]

### Datagen

In [4]:
trainpath = '../data/raw/imgs/train'
testpath = '../data/raw/imgs/testlabeled'

In [5]:
datagen = keras.preprocessing.image.ImageDataGenerator()
testdatagen = keras.preprocessing.image.ImageDataGenerator()

In [6]:
train = datagen.flow_from_directory(trainpath, \
                                    target_size=(227, 227))

Found 22424 images belonging to 10 classes.


In [7]:
test = datagen.flow_from_directory(testpath, \
                                   target_size=(227, 227))

Found 200 images belonging to 10 classes.


### Build Model

In [8]:
model = Sequential()

In [9]:
model.add(Conv2D(99, \
                 kernel_size=11, \
                 strides=4, \
                 padding='valid', \
                 activation='relu', \
                 input_shape=(227, 227, 3)))

In [10]:
model.add(MaxPool2D(3, \
                    strides=2, \
                    padding='valid'))

In [11]:
model.add(Conv2D(256, \
                 kernel_size=5, \
                 strides=1, \
                 padding='same', \
                 activation='relu'))

In [12]:
model.add(MaxPool2D(3, \
                    strides=2, \
                    padding='valid'))

In [13]:
model.add(Conv2D(384, \
                 kernel_size=3, \
                 strides=1, \
                 padding='same', \
                 activation='relu'))

In [14]:
model.add(Conv2D(384, \
                 kernel_size=3, \
                 strides=1, \
                 padding='same', \
                 activation='relu'))

In [15]:
model.add(Conv2D(256, \
                 kernel_size=3, \
                 strides=1, \
                 padding='same', \
                 activation='relu'))

In [16]:
model.add(Flatten())

In [17]:
model.add(Dense(100, activation='relu'))

In [18]:
model.add(Dense(100, activation='relu'))

In [19]:
model.add(Dense(10, activation='softmax'))

In [20]:
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

In [21]:
model.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d (Conv2D)              (None, 55, 55, 99)        36036     
_________________________________________________________________
max_pooling2d (MaxPooling2D) (None, 27, 27, 99)        0         
_________________________________________________________________
conv2d_1 (Conv2D)            (None, 27, 27, 256)       633856    
_________________________________________________________________
max_pooling2d_1 (MaxPooling2 (None, 13, 13, 256)       0         
_________________________________________________________________
conv2d_2 (Conv2D)            (None, 13, 13, 384)       885120    
_________________________________________________________________
conv2d_3 (Conv2D)            (None, 13, 13, 384)       1327488   
_________________________________________________________________
conv2d_4 (Conv2D)            (None, 13, 13, 256)       8

### Fit Model

In [22]:
model.fit(train, \
          epochs=20, \
          steps_per_epoch=40, \
          validation_data=test, \
          callbacks=callbacks_list)

  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train for 40 steps, validate for 7 steps
Epoch 1/20
40/40 [==============================] - 15s 376ms/step - loss: 17.6337 - accuracy: 0.1242 - val_loss: 2.2994 - val_accuracy: 0.1850
Epoch 2/20
40/40 [==============================] - 14s 342ms/step - loss: 2.0946 - accuracy: 0.2516 - val_loss: 1.9108 - val_accuracy: 0.2900
Epoch 3/20
40/40 [==============================] - 14s 353ms/step - loss: 1.6135 - accuracy: 0.4195 - val_loss: 1.7132 - val_accuracy: 0.3800
Epoch 4/20
40/40 [==============================] - 14s 352ms/step - loss: 1.2855 - accuracy: 0.5508 - val_loss: 1.5437 - val_accuracy: 0.4200
Epoch 5/20
40/40 [==============================] - 13s 331ms/step - loss: 1.0813 - accuracy: 0.6305 - val_loss: 1.6753 - val_accuracy: 0.4450
Epoch 6/20
40/40 [==============================] - 13s 321ms/step - loss: 0.8291 - accuracy: 0.7414 - val_loss: 2.1762 - val_accuracy: 0.4050
Epoch 7/20
40/40 [==============================] 

Definitely better, but it's only hitting ~0.46 validation accuracy